packages/arb-token-bridge-ui/src/token-bridge-sdk/CctpTransferStarter.ts

```diff
@@ -0,0 +1,140 @@
+import { prepareWriteContract, writeContract } from '@wagmi/core'
+import { constants, utils } from 'ethers'
+import { ERC20__factory } from '@arbitrum/sdk/dist/lib/abi/factories/ERC20__factory'
+
+import {
+  ApproveTokenProps,
+  BridgeTransferStarter,
+  BridgeTransferStarterProps,
+  RequiresTokenApprovalProps,
+  TransferProps,
+  TransferType
+} from './BridgeTransferStarter'
+import { formatAmount } from '../util/NumberUtils'
+import { fetchPerMessageBurnLimit, getCctpContracts } from './cctp'
+import { getChainIdFromProvider, getAddressFromSigner } from './utils'
+import { fetchErc20Allowance } from '../util/TokenUtils'
+import { TokenMessengerAbi } from '../util/cctp/TokenMessengerAbi'
+
+export class CctpTransferStarter extends BridgeTransferStarter {
+  public transferType: TransferType
+
+  constructor(props: BridgeTransferStarterProps) {
+    super(props)
+    this.transferType = 'cctp'
+  }
+
+  public requiresNativeCurrencyApproval = async () => false
+
+  public approveNativeCurrency = async () => {
+    return
+  }
+
+  public async requiresTokenApproval({
+    amount,
+    signer,
+    destinationAddress
+  }: RequiresTokenApprovalProps): Promise<boolean> {
+    const sourceChainId = await getChainIdFromProvider(this.sourceChainProvider)
+
+    const address = await getAddressFromSigner(signer)
+
+    const { usdcContractAddress } = getCctpContracts({ sourceChainId })
+
+    const recipient = destinationAddress ?? address
+    const { tokenMessengerContractAddress } = getCctpContracts({
+      sourceChainId
+    })
```

Comment:
we're calling `getCctpContracts` twice but it can be done once and you extract all the values you need